compiler/rustc_typeck/src/check/generator_interior/drop_ranges/cfg_build.rs

```diff
@@ -0,0 +1,233 @@
+    /// expressions. This method consumes a little deeper into the expression when needed.
+    fn consume_expr(&mut self, expr: &hir::Expr<'_>) {
+        debug!("consuming expr {:?}, count={}", expr.hir_id, self.expr_count);
+        let places = self
+            .places
+            .consumed
+            .get(&expr.hir_id)
+            .map_or(vec![], |places| places.iter().cloned().collect());
+        for place in places {
+            for_each_consumable(place, self.hir.find(place), |hir_id| self.record_drop(hir_id));
+        }
+    }
+
+    fn reinit_expr(&mut self, expr: &hir::Expr<'_>) {
+        if let ExprKind::Path(hir::QPath::Resolved(
+            _,
+            hir::Path { res: hir::def::Res::Local(hir_id), .. },
+        )) = expr.kind
+        {
+            let location = self.expr_count;
+            debug!("reinitializing {:?} at {}", hir_id, location);
+            self.drop_ranges.reinit_at(*hir_id, location);
+        } else {
+            debug!("reinitializing {:?} is not supported", expr);
+        }
+    }
+}
+
+impl<'tcx> Visitor<'tcx> for DropRangeVisitor<'tcx> {
+    type Map = intravisit::ErasedMap<'tcx>;
+
+    fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
+        NestedVisitorMap::None
+    }
+
+    fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) {
+        let mut reinit = None;
+        match expr.kind {
+            ExprKind::If(test, if_true, if_false) => {
+                self.visit_expr(test);
+
+                let fork = self.expr_count;
+
+                self.drop_ranges.add_control_edge(fork, self.expr_count + 1);
+                self.visit_expr(if_true);
+                let true_end = self.expr_count;
+
+                self.drop_ranges.add_control_edge(fork, self.expr_count + 1);
+                if let Some(if_false) = if_false {
+                    self.visit_expr(if_false);
+                }
+
+                self.drop_ranges.add_control_edge(true_end, self.expr_count + 1);
+            }
+            ExprKind::Assign(lhs, rhs, _) => {
+                self.visit_expr(lhs);
+                self.visit_expr(rhs);
+
+                reinit = Some(lhs);
+            }
+            ExprKind::Loop(body, ..) => {
+                let loop_begin = self.expr_count + 1;
+                self.visit_block(body);
+                self.drop_ranges.add_control_edge(self.expr_count, loop_begin);
+            }
+            ExprKind::Match(scrutinee, arms, ..) => {
+                self.visit_expr(scrutinee);
+
+                let fork = self.expr_count;
+                let arm_end_ids = arms
+                    .iter()
+                    .map(|hir::Arm { pat, body, guard, .. }| {
+                        self.drop_ranges.add_control_edge(fork, self.expr_count + 1);
+                        self.visit_pat(pat);
+                        match guard {
+                            Some(Guard::If(expr)) => self.visit_expr(expr),
```

Comment:
Example test case:

```rust
async fn main() {
    let x = vec![22_usize];
    std::mem::drop(x);
    match y() {
        true if {x = vec![]; false} => {}
        _ => { dummy().await }
    }
}

async fn dummy() { }
fn y() -> bool {true}
```

In this test, I imagine we *might* incorrectly conclude that `x` must be dropped and not re-initialized in this arm.

Reply:
Ah, good catch. I modified your example a little bit and am adding it as a test case. I like your idea of modeling match almost more like a chain of `if`s instead, so I'll do that instead.

Reply:
For patterns containing multiple alternatives the execution could also go from the guard back to the same arm.

Reply:
Good point, @tmiasko. Do you have an idea how we might observe this? I tried this:
```rust
let mut x = vec![22_usize];
std::mem::drop(x);
match false {
    true | false
        if {
            yield;
            x = vec![];
            false
        } => {}
    _ => {}
}
```
This doesn't work because I get an error on the yield saying that the borrow of `false` in the `match` is still active at the yield point. Does this mean we just can't yield in a match guard? It seems like elsewhere we can yield with borrows still active, so I'm not sure why this case doesn't work...

Here's the error message:
```
error[E0626]: borrow may still be in use when generator yields
  --> .\src\test\ui\generator\reinit-in-match-guard.rs:26:15
   |
26 |         match false {
   |               ^^^^^
...
29 |                     yield;
   |                     ----- possible yield occurs here
```

Reply:
The example could be modified to use a wildcard pattern multiple times; it doesn't introduce any borrows.